perception/crosswalk_traffic_light_estimator/src/node.cpp

```diff
@@ -219,24 +219,160 @@ void CrosswalkTrafficLightEstimatorNode::updateLastDetectedSignal(
+    const auto & elements = input_traffic_signal.second.first.elements;
+
+    if (elements.empty()) {
+      continue;
+    }
+
+    if (
+      elements.front().color == TrafficSignalElement::UNKNOWN && elements.front().confidence == 1) {
+      continue;
+    }
+
+    const auto & id = input_traffic_signal.second.first.traffic_signal_id;
+
+    if (last_colors_.count(id) == 0) {
+      std::vector<TrafficSignalAndTime> signal{input_traffic_signal.second};
+      last_colors_.insert(std::make_pair(id, signal));
+      continue;
+    }
+
+    last_colors_.at(id).push_back(input_traffic_signal.second);
+  }
+
+  std::vector<int32_t> erase_id_list;
+  for (auto & last_traffic_signal : last_colors_) {
+    const auto & id = last_traffic_signal.first;
+    for (auto it = last_traffic_signal.second.begin(); it != last_traffic_signal.second.end();) {
+      auto sig = (*it).first;
+      rclcpp::Time t = (*it).second;
+
+      // hold signal recognition results for [last_colors_hold_time_] seconds.
+      const auto time_from_last_detected = (get_clock()->now() - t).seconds();
+      if (time_from_last_detected > last_colors_hold_time_) {
+        it = last_traffic_signal.second.erase(it);
+      } else {
+        ++it;
+      }
+    }
+    if (last_traffic_signal.second.empty()) {
+      erase_id_list.emplace_back(id);
+    }
+  }
+  for (const auto id : erase_id_list) last_colors_.erase(id);
 }
 
 void CrosswalkTrafficLightEstimatorNode::setCrosswalkTrafficSignal(
-  const lanelet::ConstLanelet & crosswalk, const uint8_t color, TrafficSignalArray & msg) const
+  const lanelet::ConstLanelet & crosswalk, const uint8_t color, const TrafficSignalArray & msg,
+  TrafficSignalArray & output)
 {
   const auto tl_reg_elems = crosswalk.regulatoryElementsAs<const lanelet::TrafficLight>();
 
+  std::unordered_map<lanelet::Id, size_t> valid_id2idx_map;  // detected traffic light
+
+  for (size_t i = 0; i < msg.signals.size(); ++i) {
+    auto signal = msg.signals[i];
+    valid_id2idx_map[signal.traffic_signal_id] = i;
+  }
+
   for (const auto & tl_reg_elem : tl_reg_elems) {
-    TrafficSignal output_traffic_signal;
-    TrafficSignalElement output_traffic_signal_element;
-    output_traffic_signal_element.color = color;
-    output_traffic_signal_element.shape = TrafficSignalElement::CIRCLE;
-    output_traffic_signal_element.confidence = 1.0;
-    output_traffic_signal.elements.push_back(output_traffic_signal_element);
-    output_traffic_signal.traffic_signal_id = tl_reg_elem->id();
-    msg.signals.push_back(output_traffic_signal);
+    auto id = tl_reg_elem->id();
+    // if valid prediction exists, overwrite the estimation; else, use the estimation
+    if (valid_id2idx_map.count(id)) {
+      size_t idx = valid_id2idx_map[id];
+      auto signal = msg.signals[idx];
+      isFlashing(signal);  // check if it is flashing
+      // update output msg according to flashing and current state
+      output.signals[idx].elements[0].color = updateState(signal);
```

Comment:
```suggestion
      output.signals[idx].elements[0].color = updateFlashingState(signal);
```
not updated

Reply:
fixed in https://github.com/autowarefoundation/autoware.universe/pull/5871/commits/f3f7427e1c01e7a7c48f4e5bee06dad1b0ee2d31